rosetta/retriever/retriever_test.go

```diff
@@ -0,0 +1,1196 @@
+				return "", nil
+			},
+			TokensWithdrawnFunc: func(symbol string) (string, error) {
+				return "", nil
+			},
+		}
+		index := &mocks.Reader{
+			HeaderFunc: func(height uint64) (*flow.Header, error) {
+				assert.Equal(t, testHeight, height)
+				return testHeader, nil
+			},
+			EventsFunc: func(height uint64, types ...flow.EventType) ([]flow.Event, error) {
+				return nil, errors.New("dummy errors")
+			},
+		}
+
+		r := &Retriever{
+			validate:  validator,
+			generator: generator,
+			index:     index,
+		}
+
+		_, _, err := r.Block(testBlockID)
+		assert.Error(t, err)
+	})
+
+	t.Run("handles incorrectly-formatted indexed events", func(t *testing.T) {
+		validator := &mocks.Validator{
+			BlockFunc: func(block identifier.Block) (identifier.Block, error) { return block, nil },
+		}
+		generator := &mocks.Generator{
+			TokensDepositedFunc: func(symbol string) (string, error) {
+				return "", nil
+			},
+			TokensWithdrawnFunc: func(symbol string) (string, error) {
+				return "", nil
+			},
+		}
+		index := &mocks.Reader{
+			HeaderFunc: func(height uint64) (*flow.Header, error) {
+				assert.Equal(t, testHeight, height)
+				return testHeader, nil
+			},
+			EventsFunc: func(height uint64, types ...flow.EventType) ([]flow.Event, error) {
+				return []flow.Event{
+					{
+						Payload: []byte(`invalid_payload`),
+					},
+				}, nil
+			},
+		}
+
+		r := &Retriever{
+			validate:  validator,
+			generator: generator,
+			index:     index,
+		}
+
+		_, _, err := r.Block(testBlockID)
+		assert.Error(t, err)
+	})
+
+	t.Run("handles event with wrong amount of fields", func(t *testing.T) {
+		invalidType := &cadence.EventType{
+			Location:            utils.TestLocation,
+			QualifiedIdentifier: "test",
+			Fields: []cadence.Field{
+				{
+					Identifier: "testField",
+					Type:       cadence.UInt64Type{},
+				},
+			},
+		}
+		invalidEvent := cadence.NewEvent(
+			[]cadence.Value{
+				cadence.NewUInt64(42),
+			},
+		).WithType(invalidType)
+		invalidEventPayload := json.MustEncode(invalidEvent)
+
+		validator := &mocks.Validator{
+			BlockFunc: func(block identifier.Block) (identifier.Block, error) { return block, nil },
+		}
+		generator := &mocks.Generator{
+			TokensDepositedFunc: func(symbol string) (string, error) {
+				return "", nil
+			},
+			TokensWithdrawnFunc: func(symbol string) (string, error) {
+				return "", nil
+			},
+		}
+		index := &mocks.Reader{
+			HeaderFunc: func(height uint64) (*flow.Header, error) {
+				assert.Equal(t, testHeight, height)
+				return testHeader, nil
+			},
+			EventsFunc: func(height uint64, types ...flow.EventType) ([]flow.Event, error) {
+				return []flow.Event{
+					{
+						Payload: invalidEventPayload,
+					},
+				}, nil
+			},
+		}
+
+		r := &Retriever{
+			validate:  validator,
+			generator: generator,
+			index:     index,
+		}
+
+		_, _, err := r.Block(testBlockID)
+		assert.Error(t, err)
+	})
+
+	t.Run("handles event with missing amount field", func(t *testing.T) {
```

Comment:
I'm really not sure we need to test the event coding of the Cadence events like this. That's going a bit too far into testing a dependency (Cadence/Flow), rather than testing our own code. Overall, if the event has the correct type, it has to have the fields (unless they are declared as optionals, but that's a whole other story).

Reply:
What if we were to export the logic of converting `[]flow.Event` into rosetta objects in another package and do these tests over there? IMO still makes sense to ensure that given a proper flow event we can get the expected rosetta objects, since without that working perfectly our whole API would be broken

Reply:
I moved this logic into the `models/convert` package, under `EventsToTransactions` and moved the tests as well. Everything seems to be working fine, LMK what you think about it :)